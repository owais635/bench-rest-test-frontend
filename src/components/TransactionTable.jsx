import React from "react";
import styled from "styled-components";
import { formatDate } from "../utils/dateFormatter";
import { formatCurrency } from "../utils/currencyFormatter";

const StyledTable = styled.table`
  background: white;
  width: 100%;
  border-radius: 3px;
  border-collapse: collapse;
`;

const StyledHeader = styled.tr`
  color: ${(props) => props.theme.primaryText};
  text-align: left;

  > th {
    padding: 16px;
    font-size: small;
  }

  & > th:nth-child(4) {
    text-align: end;
  }
`;

const StyledRow = styled.tr`
  background-color: #faf9f8;
  > td {
    padding: 16px;
  }

  &:hover {
    background-color: #d3d3d340;
  }

  > td:nth-child(4) {
    text-align: end;
  }
`;

const StyledData = styled.td`
  ${(props) => (props.isMain ? " font-weight: bold;" : "")}

  color: ${({ isMain, isGain, theme }) => {
    if (isGain) return theme.primary;
    return isMain ? theme.mainText : theme.secondaryText;
  }};
`;

export default function Table({ className, total, transactions }) {
  return (
    <div className={className}>
      <StyledTable>
        <thead>
          <StyledHeader>
            <th>Date</th>
            <th>Company</th>
            <th>Account</th>
            <th>{formatCurrency(total)}</th>
          </StyledHeader>
        </thead>

        <tbody>
          {transactions.map((el, i) => {
            const props = { isGain: el.Amount > 0 };

            // We do not want to show expenses as negative
            // Expenses are shown in normal text colour, while income is shown highlighted
            const amount = el.Amount > 0 ? el.Amount : -1 * el.Amount;

            // ideally key should be a unique id and not just index.
            // Since we're not moving rows around index is ok to use
            return (
              <StyledRow index={i} key={i}>
                <StyledData {...props}>{formatDate(el.Date)}</StyledData>
                <StyledData isMain {...props}>
                  {el.Company}
                </StyledData>
                <StyledData {...props}>{el.Ledger}</StyledData>
                <StyledData isMain {...props}>
                  {formatCurrency(amount)}
                </StyledData>
              </StyledRow>
            );
          })}
        </tbody>
      </StyledTable>
    </div>
  );
}
