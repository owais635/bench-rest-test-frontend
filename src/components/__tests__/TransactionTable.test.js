import React from "react";
import renderer from "react-test-renderer";
import TransactionTable from "../TransactionTable";

jest.mock("../../utils/dateFormatter", () => ({
  formatDate: jest.fn((str) => str),
}));

const mockTransactions = [
  {
    Date: "2013-12-13",
    Ledger: "Insurance Expense",
    Amount: "-117.81",
    Company: "LONDON DRUGS 78 POSTAL VANCOUVER BC",
  },
  {
    Date: "2013-12-13",
    Ledger: "Equipment Expense",
    Amount: "-520.85",
    Company: "ECHOSIGN xxxxxxxx6744 CA xx8.80 USD @ xx0878",
  },
  {
    Date: "2013-12-13",
    Ledger: "Equipment Expense",
    Amount: "58.17",
    Company: "APPLE STORE #R280 VANCOUVER BC",
  },
];

describe("TransactionTable", () => {
  test("should render with transactions and total", () => {
    const component = renderer.create(
      <TransactionTable total={-580.49} transactions={mockTransactions} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
