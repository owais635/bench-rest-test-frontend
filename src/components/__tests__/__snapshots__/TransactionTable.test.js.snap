// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionTable should render with transactions and total 1`] = `
<div>
  <table
    className="sc-bdvvtL bwuafT"
  >
    <thead>
      <tr
        className="sc-gsDKAQ gkZXEZ"
      >
        <th>
          Date
        </th>
        <th>
          Company
        </th>
        <th>
          Account
        </th>
        <th>
          -$580.49
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="sc-dkPtRN dcSQPR"
      >
        <td
          className="sc-hKwDye hOpPvv"
        />
        <td
          className="sc-hKwDye jRYIYl"
        >
          LONDON DRUGS 78 POSTAL VANCOUVER BC
        </td>
        <td
          className="sc-hKwDye hOpPvv"
        >
          Insurance Expense
        </td>
        <td
          className="sc-hKwDye jRYIYl"
        >
          $117.81
        </td>
      </tr>
      <tr
        className="sc-dkPtRN dcSQPR"
      >
        <td
          className="sc-hKwDye hOpPvv"
        />
        <td
          className="sc-hKwDye jRYIYl"
        >
          ECHOSIGN xxxxxxxx6744 CA xx8.80 USD @ xx0878
        </td>
        <td
          className="sc-hKwDye hOpPvv"
        >
          Equipment Expense
        </td>
        <td
          className="sc-hKwDye jRYIYl"
        >
          $520.85
        </td>
      </tr>
      <tr
        className="sc-dkPtRN dcSQPR"
      >
        <td
          className="sc-hKwDye hOpPvv"
        />
        <td
          className="sc-hKwDye jRYIYl"
        >
          APPLE STORE #R280 VANCOUVER BC
        </td>
        <td
          className="sc-hKwDye hOpPvv"
        >
          Equipment Expense
        </td>
        <td
          className="sc-hKwDye jRYIYl"
        >
          $58.17
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
